File: python3/Heaps/combine_sorted_linked_lists.py
import heapq
from ds import ListNode
from typing import List

"""
Definition of ListNode:
class ListNode:
    def __init__(self, val=None, next=None):
        self.val = val
        self.next = next
"""


def combine_sorted_linked_lists(lists: List[ListNode]) -> ListNode:
    # Define a custom comparator for 'ListNode', enabling the min-heap 
    # to prioritize nodes with smaller values.
    ListNode.__lt__ = lambda self, other: self.val < other.val
    heap = []
    # Push the head of each linked list into the heap.
    for head in lists:
        if head:
            heapq.heappush(heap, head)
    # Set a dummy node to point to the head of the output linked list.
    dummy = ListNode(-1)
    # Create a pointer to iterate through the combined linked list as 
    # we add nodes to it.
    curr = dummy
    while heap:
        # Pop the node with the smallest value from the heap and add it 
        # to the output linked list.
        smallest_node = heapq.heappop(heap)
        curr.next = smallest_node
        curr = curr.next
        # Push the popped node's subsequent node to the heap.
        if smallest_node.next:
            heapq.heappush(heap, smallest_node.next)
    return dummy.next

File: java/Heaps/CombineSortedLinkedLists.java
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;

import DS.ListNode;

/*
    // Definition of ListNode:
    class ListNode {
        public int val;
        public ListNode next;
        public ListNode(int val) { this.val = val; }
    }
 */

public class CombineSortedLinkedLists {
    public ListNode combineSortedLinkedLists(List<ListNode> lists) {
        // Define a custom comparator for 'ListNode', enabling the min-heap 
        // to prioritize nodes with smaller values.
        PriorityQueue<ListNode> heap = new PriorityQueue<>(new Comparator<ListNode>() {
            @Override
            public int compare(ListNode self, ListNode other) {
                return Integer.compare(self.val, other.val);
            }
        });
        // Push the head of each linked list into the heap.
        for (ListNode head : lists) {
            if (head != null) {
                heap.offer(head);
            }
        }
        // Set a dummy node to point to the head of the output linked list.
        ListNode dummy = new ListNode(-1);
        // Create a pointer to iterate through the combined linked list as 
        // we add nodes to it.
        ListNode curr = dummy;
        while (!heap.isEmpty()) {
            // Pop the node with the smallest value from the heap and add it 
            // to the output linked list.
            ListNode smallestNode = heap.poll();
            curr.next = smallestNode;
            curr = curr.next;
            // Push the popped node's subsequent node to the heap.
            if (smallestNode.next != null) {
                heap.offer(smallestNode.next);
            }
        }
        return dummy.next;
    }
}
